### PR TITLE
CXXCBC-958: Migrate the transactions unit tests

### PR DESCRIPTION
**Pull request by avsej**
Part of [CXXCBC-945](https://jira.issues.couchbase.com/browse/CXXCBC-945). Moves the transactions unit tests off Catch2 onto the framework in `test/framework/`, into `test/unit/transactions/`.

Each leaf `SECTION` becomes its own case. Catch2 already re-runs the whole case body once per root-to-leaf section path, so the translation is exact rather than a reorganisation.

## What changes

Eight destination files: `content_as`, `get_result`, `staged_mutation`, `atr_entry`, `get_multi_transaction_id`, `get_multi_fetch`, `logging` and `utils`. They keep `LABEL unit` — `transaction` is the label for the cluster-backed suite, which this does not touch.

`content_as` asserts nothing at runtime, and that is correct: it calls a template through a null pointer inside a branch never taken, purely to force instantiation of every `content_as` overload. The regression it guards was a compile error in an overload nothing instantiated, so the binary failing to build is the failure. Catch2's `SUCCEED(...)` is dropped rather than replaced, and the comment says why the body is empty.

`logging` reached `test/utils/wait_until.hxx` for a four-line polling loop, and that header pulls the bucket, collection and user management operations — so a translation unit that opens no connection was rebuilt whenever any of them changed. The poll is now local to the file.

## What the audit found

- **`atr_entry` exercises no encode or decode.** All nine cases build an entry through its constructor and call `has_expired()`; the ATR document a server writes is parsed in `core/transactions/active_transaction_record.cxx`, which no case here reaches. Reported as a gap in decode coverage, not filled — new coverage belongs to its own ticket.
- `staged_mutation` compares `type_as_string()` against the literals `INSERT`, `REPLACE` and `REMOVE`. That is the ATR entry's `type` field, so a literal is the correct oracle for a wire constant rather than an assertion on a value the code under test produced.
- Two case names claimed more than their assertions established. `retry count in range` was one-sided, and `retry timing reasonable` named nothing; both now name the rule they hold.

## Registration

`test/unit/transactions/CMakeLists.txt` registers each file, and `test/CMakeLists.txt` gains one `add_subdirectory(unit/transactions)`.

## What this does not do

No library code. No new coverage: cases are migrated, relocated and corrected, never added.

[CXXCBC-945]: https://couchbasecloud.atlassian.net/browse/CXXCBC-945?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: test/unit/transactions/utils.cxx [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+      retry_op_exponential_backoff_timeout<void>(one_ms, ten_ms, hundred_ms, [&state] {
+        state.function();
+      });
+    },
+    "the timeout is reported rather than retried forever");
+
+  // The delays are 1+2+4+8+10+10+10+... capped at ten_ms, +/- the jitter fraction. Over a
+  // hundred_ms budget that is at most fourteen delays, so at most fifteen calls. sleep_for may
+  // sleep longer than asked, so only the upper bound is checkable.
+  assert_true(state.timings.size() < 15, "the delays are not shorter than the backoff schedule");
```
[the remark this answers is not in the document]
Comment (avsej) [reply test/unit/transactions/utils.cxx:109]: The disagreement is real and has been fixed the other way round: the assertion is correct and the comment was wrong, by one.

From retry_op_exponential_backoff_timeout in core/transactions/internal/utils.hxx, RETRY_OP_JITTER is 0.1 so the minimum multiplier is 0.9; the delay is initial_delay * (jitter() * pow(2, retries++)) and is then clamped to max_delay, so the cap is applied after jitter and is exactly ten_ms with no jitter on it. Minimum delays are 0.9, 1.8, 3.6, 7.2 ms and then ten_ms, cumulating to 93.5 ms across twelve; the thirteenth is truncated by if (now + delay > end_time) sleep_for(end_time - now), and the following if (now > end_time) break; fires. Thirteen delays, so fourteen calls. Fifteen is unreachable, because fourteen full delays would need 103.5 ms of a hundred_ms budget. Probing state.timings.size() directly returned 14 on each of eight runs.

So `< 15` admits the schedule's maximum and nothing above it, and `<= 15` would loosen a tight bound to admit a count the schedule cannot produce.

The suggested mechanism is also inverted: a faster machine cannot raise the count, because the bound is the sleep schedule against a wall-clock budget, and sleep_for overshooting lowers it. That is why only the upper bound is checkable, which the comment states.

The comment now gives the minimum-jitter schedule rather than the nominal 1+2+4+8+10..., which is what made fifteen look plausible, and names the clamp discarding the jitter.

Two further comment defects in the same file were found while checking this one and are fixed in the same commit: a cap-branch message claiming the capped delay carried jitter when the clamp replaces the jittered value, and a neighbouring case whose comment gave the exponent ceiling as 128 ms when DEFAULT_RETRY_OP_EXPONENT_CAP is 8, so it is 256 ms. The measured run time discriminates between those two: 1009.6 ms against 1023 ms nominal for a 256 ms cap, versus 767 ms for 128 ms.

File: test/unit/transactions/utils.cxx [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+    if (count == 0) {
+      assert_eq(t.count(), 0, "the first call is not delayed");
+    } else if (count < last) {
+      const auto min = min_jitter_fraction * std::pow(2, count - 1) * 1000.0;
+      if (min < 10000) {
+        assert_true(static_cast<double>(t.count()) > min,
+                    "the delay before a retry doubles, less the jitter");
+      } else {
+        assert_true(t.count() > 10000, "the delay has reached the maximum");
+      }
```
[the remark this answers is not in the document]
Comment (avsej) [reply test/unit/transactions/utils.cxx]: Applied at both sites, `>` to `>=`.

This one is accepted where the earlier bound on this file was not, and the distinction is whether the boundary value exists in the schedule. Loosening `< 15` would have admitted a call count the schedule cannot produce, since fourteen full delays need 103.5 ms of a hundred_ms budget. Here the boundary is reachable: jitter() draws from a closed [0.9, 1.1], the clamp assigns max_delay exactly, and the recorded differences are integer microseconds. A lower bound that excludes a value the implementation can produce is a bound that fails on correct behaviour, so `>=` is what the assertion means.

The messages were re-read against the new operators. "the delay before a retry doubles, less the jitter" and "the delay has reached the maximum" both describe inclusive lower bounds and needed no change; the second was already corrected in the previous round, when it claimed the capped delay still carried jitter.

[the remark this answers is not in the document]
Comment (avsej) [reply test/unit/transactions/utils.cxx]: Correction to the reply above, on the mechanism rather than the fix.

That reply said jitter() draws from a closed [0.9, 1.1] and treated the endpoint draw as the reachable path. Both are wrong. std::uniform_real_distribution yields a half-open [0.9, 1.1), so drawing exactly 0.9 is possible but has effectively zero probability, and an argument resting on it would be weak.

The reachable path is truncation. timing_differences() applies duration_cast<microseconds>, which truncates toward zero, so for the first retry any jitter in [0.9, 0.9 + 1/900) requests about 900.x microseconds, the measured delta lands just above that, and the cast yields exactly 900 against a min of 900.0. At the clamp, any measured delta in [10000, 10001) microseconds truncates to 10000. No improbable draw is needed — a sub-microsecond remainder is enough, which makes the inclusive bound necessary rather than merely tidy.

The fix is unchanged: `>=` at both sites.

One point of provenance worth recording, since it changes who owns the defect. The strict bounds were not introduced by this migration. At main both bounds in the timeout case were strict, while the sibling case `caps_the_delay_at_the_exponent_cap` already used `>= min`. They were carried over faithfully, and the sibling having had it right the whole time is the signal that should have prompted a second look.

File: test/unit/transactions/logging.cxx [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+void
+create_logger(const std::shared_ptr<capturing_sink>& sink, couchbase::core::logger::level level)
+{
+  couchbase::core::logger::configuration conf{};
+  conf.log_level = level;
+  conf.sink = sink;
+  conf.console = false;
+  couchbase::core::logger::create_file_logger(conf);
+}
```
[the remark this answers is not in the document]
Comment (avsej) [reply test/unit/transactions/logging.cxx:102]: The return value is now captured and a non-empty error fails the case with the message the logger produced, so an initialisation failure names its own cause instead of arriving as an empty sink.

File: test/unit/transactions/utils.cxx [hunk 4]
```diff
@@ -0,0 +1,474 @@
...
+      retry_op_exponential_backoff_timeout<void>(one_ms, ten_ms, hundred_ms, [&state] {
+        state.function();
+      });
+    },
+    "the timeout is reported rather than retried forever");
+
+  // sleep_for is only guaranteed to sleep for _at_least_ the time requested, so the total elapsed
+  // time has to be at least what was asked for. The timings are the times the operation was
+  // _called_; the backoff starts before the first call, so the interval before that call is added
+  // to the elapsed time rather than left out, which would let a run land just under the timeout.
```
[the remark this answers is not in the document]
Comment (avsej) [reply test/unit/transactions/utils.cxx]: Correct — the comment described a delay that does not exist. `retry_op_exponential_backoff_timeout` returns `func()` on entry and sleeps only between retries; what the adjustment compensates for is the deadline being taken before that first call. The comment now says so.

File: test/unit/transactions/utils.cxx [hunk 5]
```diff
@@ -0,0 +1,475 @@
...
+      op();
+      state.function2();
+    }
+  } catch (const retry_operation_timeout&) {
+    assert_true(state.elapsed_ms() >= hundred_ms, "the timeout is not declared before it elapses");
+    assert_true(state.timings.size() <= 15, "the delays are not shorter than the schedule");
+    return;
+  }
+  fail("the delay reports its timeout rather than blocking forever");
+}
```
[the remark this answers is not in the document]
Comment (avsej) [reply test/unit/transactions/utils.cxx:255]: Confirmed, and it was closer to failing than the report suggests. Instrumenting the assertion's own quantity — `timings.back() - timings.front()` against the 100 ms budget — over forty runs on an idle machine gave a minimum margin of **4 µs**, a mean of 52 µs and a maximum of 339 µs. Any run below the boundary truncates to 99 ms under `duration_cast<milliseconds>` and fails, so the case was one scheduling hiccup from red.

The mechanism is the one described. `exp_delay` takes its deadline on the first call and returns without sleeping, and the clamp `sleep_for(*end_time - now)` lands the last sleep on the deadline, so the recorded window is the budget minus the interval before the first timestamp — including the vector's first allocation — plus whatever `sleep_for` overshoots.

Both timeout cases now measure against the clock from a `start` taken before the first call, and neither truncates: the deadline is taken after `start` and the throw happens only once `now` is past it, so `now - start > timeout` holds by construction rather than by margin. `retry_state::elapsed_ms()` and `first_timing()` had no remaining callers and were removed.

The same case in `exponential_backoff_with_a_timeout_gives_up_when_it_elapses` carried a compensating term added in an earlier round for exactly this boundary; it is gone, since measuring the right interval needs no compensation. Nothing else in the migrated tests compares a `duration_cast<milliseconds>` elapsed time against a budget.

File: test/unit/transactions/logging.cxx [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+auto
+sink_is_empty(const std::shared_ptr<capturing_sink>& sink) -> bool
+{
+  // now, we need to be sure it is empty, and stays that way for some period of time,
+  // since async loggers don't flush immediately.   The logger is set to flush every
+  // second, so waiting 2 or 3 seems reasonable
+  if (sink->output().empty()) {
+    std::this_thread::sleep_for(std::chrono::seconds(2));
```
[the remark this answers is not in the document]
Comment (avsej) [reply test/unit/transactions/logging.cxx]: Replaced, and the case is both faster and stronger for it. Each negative case now logs a message the filter accepts after the one it drops and waits for that sentinel: the queue is ordered, so once the sentinel is in the sink a dropped message would already be there had it been accepted. A worker delayed past a deadline can no longer let a broken filter pass, which was the real defect in the fixed wait rather than its length.

The helper it replaces is gone, both cases dropped `timeout::slow`, and they now cost one poll interval each instead of two seconds:

```
before   a_custom_sink_respects_the_configured_level          2000ms
after    a_custom_sink_respects_the_configured_level           103ms
```

File: test/unit/transactions/utils.cxx [hunk 7]
```diff
@@ -0,0 +1,465 @@
...
+  // Minimum jitter maximises the count: the delays are 0.9, 1.8, 3.6, 7.2 ms and are then
+  // clamped to ten_ms with the jitter discarded, so 13.5 ms of the hundred_ms budget buys eight
+  // more full delays and a ninth truncated against the deadline. Thirteen delays, so fourteen
+  // calls, and fifteen is out of reach. sleep_for may overshoot, which lowers the count, so only
+  // the upper bound is checkable.
+  assert_true(state.timings.size() < 15, "the delays are not shorter than the backoff schedule");
+}
+
+void
+exponential_backoff_with_a_timeout_grows_the_delay_between_retries([[maybe_unused]] context& ctx)
```
[the remark this answers is not in the document]
Comment (avsej) [reply test/unit/transactions/utils.cxx]: Correct, and the same holds one case further down. A constant delay at the clamp satisfies every bound here, so the name claimed growth the assertions never checked.

Renamed to `exponential_backoff_with_a_timeout_never_retries_faster_than_the_schedule`, with the comment saying why the floor is all that is checkable: `sleep_for` may overshoot by any amount, so an upper bound on an observed delay is a flake rather than a specification. That is also why the growth is not asserted instead of renamed.

File: test/unit/transactions/utils.cxx [hunk 8]
```diff
@@ -0,0 +1,465 @@
...
+        state.function();
+      });
+    },
+    "exhausting the attempts is reported rather than retried forever");
+
+  assert_eq(state.timings.size(), std::size_t{ 1 }, "a retry budget of zero still calls once");
+}
+
+void
+exponential_backoff_caps_the_delay_at_the_exponent_cap([[maybe_unused]] context& ctx)
```
[the remark this answers is not in the document]
Comment (avsej) [reply test/unit/transactions/utils.cxx]: Correct on both counts, and the second is the stronger one. With `DEFAULT_RETRY_OP_EXPONENT_CAP` at 8 and ten retries the intervals are 1, 2, 4, ..., 128, 256, 256 ms: the only interval where a capped and an uncapped run differ is the last, which `count < last` excludes, and a lower bound of 0.9 x 256 would admit an uncapped 512 anyway. The case never observed the cap.

Renamed to `exponential_backoff_never_retries_faster_than_the_capped_schedule`, and the comment now records that an uncapped run would clear every floor below it. Adding an upper bound would make the case a timing flake, so the rule it holds is stated rather than strengthened.

**Review by brett19 — APPROVED** [review]

**Comment by avsej** [comment]
The finding was correct at the commit it reviewed, and no longer applies to the head.

That review examined `44eb8deb`, a merge of `main` into this branch. Its resolution of `test/CMakeLists.txt` kept ten `unit_test(...)` registrations whose sources had been deleted: the four JSON suites, which `add_subdirectory(unit/json)` already registers, and the six transaction suites this branch moves. Configure fails on that commit for exactly the reason given.

The head is now `abc15109`, which replaces that merge with a rebase onto `main`. Every `unit_test(...)` registration there resolves to an existing `test/test_unit_*.cxx`, and configure, build, the `unit` suite and clang-format are checked per commit across the branch.

One part of the finding does not hold at either commit: the `transaction_test(...)` calls further down are a different macro, building `test/test_transaction_*.cxx`. Those are the integration transaction suites, which this branch does not touch, so they are not stale registrations.

The underlying hazard is what the merge introduced. Sibling branches in this series delete `test_unit_*` sources that this one keeps, so resolving the shared `test/CMakeLists.txt` in favour of either side leaves registrations that name files the other side removed. Each branch is rebased on `main` rather than merged, and re-verified per commit, so the resolution is never carried in a commit.

File: test/unit/transactions/utils.cxx [hunk 9]
```diff
@@ -0,0 +1,474 @@
...
+  assert_eq(public_result.id(), f.key, "the key survives the conversion");
+  assert_true(public_result.content_as<tao::json::value>() == f.content,
+              "the content survives the conversion");
+  assert_eq(core_result.collection(), f.collection, "the collection is unchanged by converting");
+  assert_eq(core_result.bucket(), f.bucket, "the bucket is unchanged by converting");
+  assert_eq(core_result.scope(), f.scope, "the scope is unchanged by converting");
+  assert_true(core_result.cas() == f.cas, "the CAS is unchanged by converting");
+  assert_eq(core_result.key(), f.key, "the key is unchanged by converting");
+  // The content is _moved_ into the public result, so core_result.content() is left valid but
+  // holding anything; nothing below may read it.
```
[the remark this answers is not in the document]
Comment (avsej) [reply test/unit/transactions/utils.cxx]: Reworded: the moved-from result is left in a valid but unspecified state, nothing below reads it, and nothing in the case asserts what it holds.

The suppressed note on `atr_entry.cxx` is also addressed, and it was right about more than the wording. `elapsed > ttl + margin` and `elapsed - margin > ttl` are the same inequality, so no assertion could distinguish the two formulations the message claimed to separate. The message now states the observable behaviour instead: a margin wider than the overrun keeps an entry past its TTL unexpired.
